android/app/src/main/java/com/google/samples/cronet_sample/MainActivity.java

```diff
@@ -27,47 +27,62 @@
 import androidx.recyclerview.widget.RecyclerView;
 import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;
 import com.google.samples.cronet_sample.data.ImageRepository;
+import java.io.File;
+import java.io.IOException;
 import java.util.concurrent.atomic.AtomicLong;
+import org.chromium.net.CronetEngine;
 
 public class MainActivity extends AppCompatActivity {
 
     private static final String TAG = "MainActivity";
     private SwipeRefreshLayout swipeRefreshLayout;
-    private AtomicLong cronetLatency = new AtomicLong();
+    private final AtomicLong cronetLatency = new AtomicLong();
     private long totalLatency;
     private long numberOfImages;
-    private ViewAdapter viewAdapter;
+    public static CronetEngine cronetEngine;
 
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
+
+        // set up cronet engine and it's logging mechanism
+        Thread.setDefaultUncaughtExceptionHandler(new ExceptionHandler());
+        setCronetEngine();
+        startNetLog();
+
         setContentView(R.layout.images_activity);
         setUpToolbar();
-        swipeRefreshLayout = (SwipeRefreshLayout) findViewById(R.id.images_activity_layout);
+        swipeRefreshLayout = findViewById(R.id.images_activity_layout);
         swipeRefreshLayout.setOnRefreshListener(() -> loadItems());
         loadItems();
     }
 
+    @Override
+    protected void onDestroy() {
+        super.onDestroy();
+
+        // ensure logging is stopped properly when activity is ended, either by system or
+        // by clicking the back button
+        stopNetLog();
+    }
+
     private void loadItems() {
         numberOfImages = 0;
 
-        RecyclerView cronetView = (RecyclerView) findViewById(R.id.images_view);
+        RecyclerView cronetView = findViewById(R.id.images_view);
+
         GridLayoutManager gridLayoutManager =
                 new GridLayoutManager(this, 2);
 
-        viewAdapter = new ViewAdapter(this);
-        // In order to enable Netlog, a Cronet logging system, enable write permissions.
-        // Find more info about Netlog here:
-        // https://www.chromium.org/developers/design-documents/network-stack/netlog
-        enableWritingPermissionForLogging();
-
         cronetView.setLayoutManager(gridLayoutManager);
-        cronetView.setAdapter(viewAdapter);
+        cronetView.setAdapter(new ViewAdapter(this));
         cronetView.setItemAnimator(new DefaultItemAnimator());
         onItemsLoadComplete();
 
     }
 
+    // External storage access is not allowed for android api level >= 30
+    // See doc: https://developer.android.com/about/versions/11/privacy/storage
     private void enableWritingPermissionForLogging() {
```

Comment:
Can we delete the method then?

Reply:
I left it there incase people wanted to see how to enable write access for older api levels.